e2etests/node_scan_rhcos_test.go

```diff
@@ -0,0 +1,204 @@
+				// We conduct a spot-checking here - more vulns can be returned from scanner for libksba and tar,
+				// but we care only about the selected one as it is sufficient for this test case.
+				// (We do not test that the set of vulns is complete, we test that the API returns any vulns if expected).
+				// We use 'equals' assertion if 0 vulns are expected.
+				Features: []*v1.Feature{
+					{
+						Name:            "libksba",
+						Version:         "1.3.5-7.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{vulnLibksba},
+					},
+					{
+						Name:            "tar",
+						Version:         "1.27.1.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{vulnTar},
+					},
+					{
+						Name:            "grep",
+						Version:         "3.1-6.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{},
+					},
+				},
+				NodeNotes: nil,
+			},
+		},
+		{
+			name:    "Uncertified scan is unsupported for RHCOS and returns no features",
+			request: buildRequest([]v1.Note{v1.Note_CERTIFIED_RHEL_SCAN_UNAVAILABLE}),
+			expectedResponse: &v1.GetNodeVulnerabilitiesResponse{
+				Features:  []*v1.Feature{},
+				NodeNotes: nil,
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			c := c
+			gotResponse, err := client.GetNodeVulnerabilities(context.Background(), c.request)
+			require.NoError(t, err)
+			assert.Len(t, gotResponse.GetFeatures(), len(c.expectedResponse.GetFeatures()), "Unexpected number of features") // unusual got-expected order of Len
+			assert.Len(t, gotResponse.GetNodeNotes(), len(c.expectedResponse.GetNodeNotes()))
+
+			for _, expectedFeat := range c.expectedResponse.GetFeatures() {
+				foundFeat := findFeat(expectedFeat.GetName(), expectedFeat.GetVersion(), gotResponse.GetFeatures())
+				assert.NotNil(t, foundFeat, "Expected to find feature '%s:%s'", expectedFeat.GetName(), expectedFeat.GetVersion())
+				if foundFeat == nil {
+					continue
+				}
+				// when 0 vulns are expected, then use stronger assertion, because empty set is a subset of any set
+				if len(expectedFeat.GetVulnerabilities()) == 0 {
+					assert.Len(t, foundFeat.GetVulnerabilities(), 0, "Expected to find 0 vulnerabilities for feature '%s:%s'", expectedFeat.GetName(), expectedFeat.GetVersion())
+				} else {
+					assertIsSubset(t, expectedFeat.GetVulnerabilities(), foundFeat.GetVulnerabilities())
+				}
+			}
+		})
+	}
+}
+
+func findFeat(name, version string, set []*v1.Feature) *v1.Feature {
+	for _, gotFeat := range set {
+		if gotFeat.GetName() == name && gotFeat.GetVersion() == version {
+			return gotFeat
+		}
+	}
+	return nil
+}
+
+// assertIsSubset asserts that every element of 'subset' exists in 'set'
+func assertIsSubset(t *testing.T, subset, set []*v1.Vulnerability) {
+	assert.GreaterOrEqual(t, len(set), len(subset), "Expected to find at least %d vulnerabilities", len(subset))
+	// Prune last modified time
+	for _, v := range set {
+		if v == nil {
+			continue
+		}
```

Comment:
Hm. Usually a list (slice) should not contain `nil` entries because "absent" (or "empty") values can be just not included in the list. If this list doesn't actually contain `nil`-s, than this condition can be dropped.
```suggestion
```

Reply:
I can remove this as this is only a test code, but an API response in shape of `[]*v1.Vulnerability{nil, nil}` can theoretically happen and then the test will panic instead of failing. 

Theoretically, the goproto generated code should not allow to send `[]*v1.Vulnerability{nil, nil}` as a reply... and it does that by panicking. I hope the scanner will never return such value.

Reply:
Panic in test is perfectly valid outcome. Our life will be impossibly hard if we have to check slices/lists for `nil` elements.